# Tokyo Apartments Price Prediction

# Introduction

The aim of this project is to analyze the pricing of apartments in Tokyo. The reason why I chose to do a project on this is because my parents are living in Tokyo, and the insights I gain through this project will be useful for me in case I ever need to find an apartment to rent to go take care of them.
<br><br>
I will start from fetching the data myself by web scraping. To gather the data, I will use Selenium to web scrape the data from apartments.gaijinpot.com, which is an apartment listing website catered towards expats. Then, I will go through the entire ETL process by cleaning the data and then loading it to a PostgreSQL database.
<br><br>
After the data has been cleaned, I will do analysis on Tableau to figure out which types of information explain apartment pricing. Then I will construct a dashboard to condense relevant information into one visualization. 
<br> <br>
In this notebook, I will be going over the ETL process for this project.


# Part One: Data Gathering (Extract)

The data will be gathered from https://apartments.gaijinpot.com/en/rent/tokyo. I will use Selenium, as it has the capability to automate scrolling and clicking to different pages.

In [17]:
# Load required libraries
import pandas as pd
import numpy as np
import requests
from selenium import webdriver
import re
import warnings
warnings.filterwarnings('ignore')
warnings.simplefilter('ignore')

In [89]:
# Step 1: Specify which browser to use on Selenium and where the driver for it is located
from selenium import webdriver
PATH = 'C:\Program Files (x86)\chromedriver.exe'
driver = webdriver.Chrome('C:\Program Files (x86)\chromedriver.exe')

In [90]:
# Step 2: Open the yahoo finance website using the driver.get() method
url = 'https://apartments.gaijinpot.com/en/rent/tokyo'
driver.get(url)

In [29]:
# Step 3: Scrape the HTML source code of the table where the data is located

# The listings variable contains each apartment listing on the webpage

listings = driver.find_elements_by_class_name("listing-body")



In [82]:
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time

In [149]:
# Add "Not Available" to lists that did not scrape info from the listing as the listing had missing information
def not_available_padding(bigger_list, smaller_list):
    if len(bigger_list) > len(smaller_list):
        while len(bigger_list) > len(smaller_list):
            smaller_list.append("Not Available")

The next algorithm scrapes all the listings in the website. The algorithm can be explained in n parts:
<br><br>
1. The list that holds the listing information is initialized 
2. Selenium finds all the listings on the current web page
3. For each listing it finds, the algorithm checks how many pieces of information it found for the listing (out of 9 possible)
4. For each information, it goes through checks to see which type of information it is by checking for key words. For example, if the information has the string "Available", then the information will be appended to the availability list.
5. After the algorithm checks every key word for every information found, then the algorithm checks if there were any information missing by comparing the lengths of the different lists that hold the scraped information. For example, if a listing was missing information about the nearest station, then the length of the nearest station list is smaller than that of other lists. If the algorithm finds missing information, the list with the missing information will be appended with a "Not Available" string.
6. The algorithm looks for a button to move to the next page of listings and clicks on it
7. Repeat steps 2 to 6 until the button to move to the next page is not found (when Selenium reached the last page of listings). 

<br><br>
This is a slow algorithm, but is necessary to ensure that all scraped info will be allocated to the correct lists. Without the checks for key words, incorrect information may end up in different lists when listing information is missing. Through trial and error, I've found that the listing always has a monthly cost information, while it may sometimes lack flooring, year built, or nearest station information.

In [166]:
# Automates finding listings, scraping listings, and clicking to the next page until it reaches the last page

listing_title = []
monthly_costs = []
availability = []
size = []
deposit = []
key_price = []
floor = []
year_built = []
nearest_station = []


while True:
    try:
        
        listings = driver.find_elements_by_class_name("listing-body") # Find all listings within the current page
        
        # An algorithm to capture all listings in the proper lists
        for listing in listings:
            for i in range(len(listing.find_elements_by_class_name("listing-item"))): # For however many info about the listing was found per listing
                if '\n' in listing.find_elements_by_class_name("listing-item")[i].text:
                    listing_title.append(listing.find_elements_by_class_name("listing-item")[i].text) # Append to listing title
                    continue
                elif 'Monthly Costs' in listing.find_elements_by_class_name("listing-item")[i].text: 
                    monthly_costs.append(listing.find_elements_by_class_name("listing-item")[i].text) # Append monthly costs if info is correct
                    
                elif 'Available' in listing.find_elements_by_class_name("listing-item")[i].text: 
                    availability.append(listing.find_elements_by_class_name("listing-item")[i].text) # Add availability 
                    
                elif 'Size' in listing.find_elements_by_class_name("listing-item")[i].text: 
                    size.append(listing.find_elements_by_class_name("listing-item")[i].text) # Add size
                    
                elif 'Deposit' in listing.find_elements_by_class_name("listing-item")[i].text: 
                    deposit.append(listing.find_elements_by_class_name("listing-item")[i].text) # Add deposit
                    
                elif 'Key Money' in listing.find_elements_by_class_name("listing-item")[i].text: 
                    key_price.append(listing.find_elements_by_class_name("listing-item")[i].text) # Add key price
                    
                elif 'Floor' in listing.find_elements_by_class_name("listing-item")[i].text: 
                    floor.append(listing.find_elements_by_class_name("listing-item")[i].text) # Add floor information
                    
                elif 'Year Built' in listing.find_elements_by_class_name("listing-item")[i].text: 
                    year_built.append(listing.find_elements_by_class_name("listing-item")[i].text) # Add year built information
                    
                elif 'Nearest Station' in listing.find_elements_by_class_name("listing-item")[i].text: 
                    nearest_station.append(listing.find_elements_by_class_name("listing-item")[i].text) # Add nearest station information
            
            # Check if there were any missing information by comparing length of lists. Nearest station and Floor gets empty info
            not_available_padding(monthly_costs, floor)
            not_available_padding(monthly_costs, year_built)
            not_available_padding(monthly_costs, nearest_station)

                
    
        element = WebDriverWait(driver, 0.01).until( 
                EC.presence_of_element_located((By.CLASS_NAME, "pagination-next"))
        ) # Find the next button to click
        element.click()
        

    except: # Breaks the look when the scraper reaches the last page
        print("Done")
        break
        

Done


In [169]:
# Save the scraped data 
dataset = pd.DataFrame(zip(listing_title, monthly_costs, availability, size, 
                           deposit, key_price, floor, year_built, nearest_station), 
                       columns = ["Listing Title", "Monthly Costs", "Availability", "Size", 
                                 "Deposit", "Key Price", "Floor", "Year Built", "Nearest Station"])

# Save the data into a csv
dataset.to_csv(r"C:\Users\taked\Desktop\DataScience Projects\Non-Kaggle Datasets\Tokyo Apartments Project\scraped_data.csv")

Now that the data has been gathered, I will be cleaning the data for later analysis.

# Part Two: Data Cleaning (Transform)

In [18]:
# Reading the dataset
dataset = pd.read_csv(r"C:\Users\taked\Desktop\DataScience Projects\Non-Kaggle Datasets\Tokyo Apartments Project\scraped_data.csv",
                     index_col = False)
warnings.simplefilter(action='ignore', category=FutureWarning)

In [19]:
# Looking at the general format of the data
dataset.head(2)

Listing Title  \
0  PRIVATE GUESTHOUSE\nIN SHIMOTAKAIDO\nSUGINAMI-...   
1       1K APARTMENT\nIN HOMMACHI\nSHIBUYA-KU, TOKYO   

                       Monthly Costs   Availability           Size  \
0             Monthly Costs ¥140,000  Available Now  Size 12.00 m²   
1  Monthly Costs ¥174,000 negotiable  Available Now  Size 37.10 m²   

      Deposit     Key Price         Floor       Year Built  \
0  Deposit ¥0  Key Money ¥0  Floor 2 / 8F  Year Built 2019   
1  Deposit ¥0  Key Money ¥0  Floor 1 / 4F  Year Built 2005   

                                     Nearest Station  
0  Nearest Station Kamikitazawa Station (5 min. w...  
1     Nearest Station Hatsudai Station (9 min. walk)

In [20]:
# Removing duplicated rows
no_dup = dataset.drop_duplicates(keep = "first")

In [21]:
# Checking if the removal of duplication was properly executed
dataset[dataset["Listing Title"] == 'PRIVATE GUESTHOUSE\nIN SHIMOTAKAIDO\nSUGINAMI-KU, TOKYO'].head(3)

Listing Title           Monthly Costs  \
0   PRIVATE GUESTHOUSE\nIN SHIMOTAKAIDO\nSUGINAMI-...  Monthly Costs ¥140,000   
16  PRIVATE GUESTHOUSE\nIN SHIMOTAKAIDO\nSUGINAMI-...  Monthly Costs ¥140,000   
32  PRIVATE GUESTHOUSE\nIN SHIMOTAKAIDO\nSUGINAMI-...  Monthly Costs ¥140,000   

     Availability           Size     Deposit     Key Price         Floor  \
0   Available Now  Size 12.00 m²  Deposit ¥0  Key Money ¥0  Floor 2 / 8F   
16  Available Now  Size 12.00 m²  Deposit ¥0  Key Money ¥0  Floor 2 / 8F   
32  Available Now  Size 12.00 m²  Deposit ¥0  Key Money ¥0  Floor 2 / 8F   

         Year Built                                    Nearest Station  
0   Year Built 2019  Nearest Station Kamikitazawa Station (5 min. w...  
16  Year Built 2019  Nearest Station Kamikitazawa Station (5 min. w...  
32  Year Built 2019  Nearest Station Kamikitazawa Station (5 min. w...

In [22]:
# Checked
no_dup[no_dup["Listing Title"] == 'PRIVATE GUESTHOUSE\nIN SHIMOTAKAIDO\nSUGINAMI-KU, TOKYO']

Listing Title           Monthly Costs  \
0  PRIVATE GUESTHOUSE\nIN SHIMOTAKAIDO\nSUGINAMI-...  Monthly Costs ¥140,000   

    Availability           Size     Deposit     Key Price         Floor  \
0  Available Now  Size 12.00 m²  Deposit ¥0  Key Money ¥0  Floor 2 / 8F   

        Year Built                                    Nearest Station  
0  Year Built 2019  Nearest Station Kamikitazawa Station (5 min. w...

It seems that every column needs to be cleaned to be useful as features. Therefore, I will tackle the columns one by one.

<br>
Listing Title Column

In [102]:
# Observing the strings
no_dup["Listing Title"].values[0:4]

array(['PRIVATE GUESTHOUSE\nIN SHIMOTAKAIDO\nSUGINAMI-KU, TOKYO',
       '1K APARTMENT\nIN HOMMACHI\nSHIBUYA-KU, TOKYO',
       '1LDK APARTMENT\nIN WAKABA\nSHINJUKU-KU, TOKYO',
       '1LDK APARTMENT\nIN HOMMACHI\nSHIBUYA-KU, TOKYO'], dtype=object)

The string that maeks up the listing title is separated by "\n" which can be used to an advantage. The format of this string follows: "[room detail]\nIN [city name]\n[ward name], TOKYO"
<br>
One strategy I can do is to first convert all "\n" into commas, then separate the string by commas, take the first portion of the separated string as the room detail, the second portion as the city name, and the third as the ward name.

In [23]:
# Cleaning the Listing Title column


# Replacing '/n' string with a comma
no_dup["Listing Title"] = no_dup["Listing Title"].str.replace('\n', ',')

# Copying the listing title column to a list
listing_titles = no_dup["Listing Title"].to_list()

# Algorithm to separate and store appropriate information from listing title to different categories of listing information

room_detail = []
city_name = []
ward_name = []
for item in listing_titles:
    splitted_info = item.split(',')
    room_detail.append(splitted_info[0])
    city_name.append(splitted_info[1])
    ward_name.append(splitted_info[2])

# Updating dataframe with new data
no_dup["Room Details"] = room_detail
no_dup["City"] = city_name
no_dup["Ward"] = ward_name

# Cleaning the city & ward columns by removing unnecessary string
no_dup["City"] = no_dup["City"].str.replace("IN ", "")
no_dup["Ward"] = no_dup["Ward"].str.replace("-KU", "")
no_dup["Ward"] = no_dup["Ward"].str.replace("-SHI", "")

# Drop the original column
no_dup.drop("Listing Title", axis = 1, inplace = True)

<br>
Monthly Costs column

In [24]:
no_dup.head(2)

Monthly Costs   Availability           Size  \
0             Monthly Costs ¥140,000  Available Now  Size 12.00 m²   
1  Monthly Costs ¥174,000 negotiable  Available Now  Size 37.10 m²   

      Deposit     Key Price         Floor       Year Built  \
0  Deposit ¥0  Key Money ¥0  Floor 2 / 8F  Year Built 2019   
1  Deposit ¥0  Key Money ¥0  Floor 1 / 4F  Year Built 2005   

                                     Nearest Station        Room Details  \
0  Nearest Station Kamikitazawa Station (5 min. w...  PRIVATE GUESTHOUSE   
1     Nearest Station Hatsudai Station (9 min. walk)        1K APARTMENT   

           City      Ward  
0  SHIMOTAKAIDO  SUGINAMI  
1      HOMMACHI   SHIBUYA

It seems that some monthly rents are negotiable, while some are not. Therefore, my plan is to first create a binary categorical variable of if the rent expense is negotiable or not. Then, I will extract numbers from each string value in the Monthly Costs column using regex.

In [25]:
# Cleaning the Monthly Costs column

# Extracting negotiation information
no_dup["Negotiable"] = np.where(no_dup["Monthly Costs"].str.contains("negotiable"), "Yes", "No")

# Extracting monthly costs 

monthly_rents_in_yen = []
rent_pattern = r"\d+,\d+"
rent = no_dup["Monthly Costs"].to_list()
for item in rent:
    match = re.search(rent_pattern,item).group(0)       # Find regex match based on the pattern
    split_rent_str = match.split(',')                   # Preprocess the matched string by splitting the rent price by comma
    rent_cleaned = float(''.join(split_rent_str))      # Attaching the split strings together and converting to float
    monthly_rents_in_yen.append(rent_cleaned)           # Appending the cleaned monthly rent 
    
# Updating dataframe with new data
no_dup["Monthly Rent"] = monthly_rents_in_yen

# Qualitative check 
no_dup[["Monthly Costs", "Monthly Rent"]]

Monthly Costs  Monthly Rent
0                 Monthly Costs ¥140,000      140000.0
1      Monthly Costs ¥174,000 negotiable      174000.0
2      Monthly Costs ¥325,000 negotiable      325000.0
3      Monthly Costs ¥177,000 negotiable      177000.0
4                 Monthly Costs ¥120,000      120000.0
...                                  ...           ...
17353              Monthly Costs ¥61,000       61000.0
17354              Monthly Costs ¥22,000       22000.0
17355             Monthly Costs ¥220,000      220000.0
17356             Monthly Costs ¥320,000      320000.0
17357             Monthly Costs ¥200,000      200000.0

[15944 rows x 2 columns]

Although the data cleaning process was successful, I see an outlier with a monthly cost of 22,000 yen, which is roughly only $220. I will keep this in mind when doing exploratory data analysis, so that I will investigate further as to why that listing has such a low rent.

In [26]:
# Drop the original column
no_dup.drop("Monthly Costs", axis = 1, inplace = True)

<br>
Availability

I am going to transform the Availability column into a binary column, whether the apartment is available now or not.

In [13]:
no_dup.head(3)

Availability           Size     Deposit     Key Price  \
0               Available Now  Size 12.00 m²  Deposit ¥0  Key Money ¥0   
1               Available Now  Size 37.10 m²  Deposit ¥0  Key Money ¥0   
2  Available From Jul 1, 2021  Size 44.81 m²  Deposit ¥0  Key Money ¥0   

          Floor       Year Built  \
0  Floor 2 / 8F  Year Built 2019   
1  Floor 1 / 4F  Year Built 2005   
2  Floor 5 / 8F  Year Built 2015   

                                     Nearest Station        Room Details  \
0  Nearest Station Kamikitazawa Station (5 min. w...  PRIVATE GUESTHOUSE   
1     Nearest Station Hatsudai Station (9 min. walk)        1K APARTMENT   
2      Nearest Station Yotsuya Station (7 min. walk)      1LDK APARTMENT   

           City      Ward Negotiable  Monthly Rent  
0  SHIMOTAKAIDO  SUGINAMI         No      140000.0  
1      HOMMACHI   SHIBUYA        Yes      174000.0  
2        WAKABA  SHINJUKU        Yes      325000.0

In [27]:
# Create new column "Available Now?" that takes the values "Yes" or "No" depending on availability

no_dup["Available Now?"] = np.where(no_dup["Availability"].str.contains("Now"), "Yes", "No")

In [25]:
no_dup

Availability           Size           Deposit  \
0                   Available Now  Size 12.00 m²        Deposit ¥0   
1                   Available Now  Size 37.10 m²        Deposit ¥0   
2      Available From Jul 1, 2021  Size 44.81 m²        Deposit ¥0   
3                   Available Now  Size 57.00 m²        Deposit ¥0   
4                   Available Now  Size 23.00 m²  Deposit ¥120,000   
...                           ...            ...               ...   
17353               Available Now  Size 10.00 m²   Deposit ¥20,000   
17354               Available Now   Size 2.20 m²        Deposit ¥0   
17355               Available Now  Size 36.32 m²  Deposit ¥205,000   
17356               Available Now  Size 59.91 m²  Deposit ¥320,000   
17357               Available Now  Size 70.16 m²  Deposit ¥195,000   

                Key Price         Floor       Year Built  \
0            Key Money ¥0  Floor 2 / 8F  Year Built 2019   
1            Key Money ¥0  Floor 1 / 4F  Year Built 2005   
2            Key Money ¥0  Floor 5 / 8F  Year Built 2015   
3            Key Money ¥0  Floor 1 / 3F  Year Built 1994   
4            Key Money ¥0  Floor 1 / 3F  Year Built 2020   
...                   ...           ...              ...   
17353        Key Money ¥0  Floor 1 / 2F  Year Built 2006   
17354        Key Money ¥0  Floor 1 / 5F  Year Built 2016   
17355  Key Money ¥205,000  Floor 7 / 7F  Year Built 2015   
17356  Key Money ¥320,000  Floor 7 / 7F  Year Built 2013   
17357  Key Money ¥195,000  Floor 2 / 3F  Year Built 2018   

                                         Nearest Station        Room Details  \
0      Nearest Station Kamikitazawa Station (5 min. w...  PRIVATE GUESTHOUSE   
1         Nearest Station Hatsudai Station (9 min. walk)        1K APARTMENT   
2          Nearest Station Yotsuya Station (7 min. walk)      1LDK APARTMENT   
3         Nearest Station Hatsudai Station (9 min. walk)      1LDK APARTMENT   
4         Nearest Station Asakusa Station (10 min. walk)        1R APARTMENT   
...                                                  ...                 ...   
17353  Nearest Station Gakugei Daigaku Station (3 min...  PRIVATE GUESTHOUSE   
17354  Nearest Station Gakushuinshita Station (4 min....              OFFICE   
17355        Nearest Station Hiroo Station (9 min. walk)       1DK APARTMENT   
17356   Nearest Station Daikanyama Station (4 min. walk)      2LDK APARTMENT   
17357    Nearest Station Tachikawa Station (7 min. walk)      2LDK APARTMENT   

                City       Ward Negotiable  Monthly Rent Available Now?  
0       SHIMOTAKAIDO   SUGINAMI         No      140000.0            Yes  
1           HOMMACHI    SHIBUYA        Yes      174000.0            Yes  
2             WAKABA   SHINJUKU        Yes      325000.0             No  
3           HOMMACHI    SHIBUYA        Yes      177000.0            Yes  
4            ASAKUSA      TAITO         No      120000.0            Yes  
...              ...        ...        ...           ...            ...  
17353        TAKABAN     MEGURO         No       61000.0            Yes  
17354         TAKADA    TOSHIMA         No       22000.0            Yes  
17355          EBISU    SHIBUYA         No      220000.0            Yes  
17356  DAIKANYAMACHO    SHIBUYA         No      320000.0            Yes  
17357   SHIBASAKICHO  TACHIKAWA         No      200000.0            Yes  

[15944 rows x 13 columns]

In [28]:
# Drop the original column
no_dup.drop("Availability", axis = 1, inplace = True)

<br>
Size 

The cleaning for this column will be very straight forward; I will just extract the float within the values of this column using regular expression and create a new column with it. I will need to keep in mind that the size is in squared meters.
<br><br>
Because the cleaning of all subsequent columns require some sort of regular expressions, I will define a helper function to expedite the process.

In [29]:
# Helper function to extract information using regular expression. 
# If the information to be extracted is in yen, the script will run additional cleaning
def regex_extract(orig_col, new_col, regex_pattern, dtype, yen = False):
    cleaned_data = []
    orig_data = no_dup[orig_col].to_list()
    for item in orig_data:
        match = re.search(regex_pattern,item).group(0)       # Find regex match based on the pattern
        if yen: #Additional cleaning if rent
            split_str = match.split(',')                   # Preprocess the matched string by splitting the rent price by comma
            match =''.join(split_str)     # Attaching the split strings together 
        if dtype == "int":
            match = int(match)
        elif dtype == "float":
            match = float(match)
        
        cleaned_data.append(match)           # Appending the cleaned monthly rent 

    # Updating dataframe with new data
    no_dup[new_col] = cleaned_data

    # Qualitative check 
    print(no_dup[[orig_col, new_col]])


In [30]:
# CLeaning the Size column

regex_extract("Size", "Size in squared meters", r"\d+.\d+", dtype = "float")

                Size  Size in squared meters
0      Size 12.00 m²                   12.00
1      Size 37.10 m²                   37.10
2      Size 44.81 m²                   44.81
3      Size 57.00 m²                   57.00
4      Size 23.00 m²                   23.00
...              ...                     ...
17353  Size 10.00 m²                   10.00
17354   Size 2.20 m²                    2.20
17355  Size 36.32 m²                   36.32
17356  Size 59.91 m²                   59.91
17357  Size 70.16 m²                   70.16

[15944 rows x 2 columns]


In [31]:
# Drop the original column
no_dup.drop("Size", axis = 1, inplace = True)

<br>
Deposit
<br>
I will first use the helper function defined above to extract the yen information. Then, I will make another column, which will be a binary one,
to indicate whether the listing required a deposit or not.

In [32]:
# Cleaning the Deposit column

regex_extract("Deposit", "Deposit in Yen", r"\d+(.\d+)?", dtype = "int", yen = True)

no_dup["Deposit Required?"] = np.where(no_dup["Deposit in Yen"] == 0, "No", "Yes")

                Deposit  Deposit in Yen
0            Deposit ¥0               0
1            Deposit ¥0               0
2            Deposit ¥0               0
3            Deposit ¥0               0
4      Deposit ¥120,000          120000
...                 ...             ...
17353   Deposit ¥20,000           20000
17354        Deposit ¥0               0
17355  Deposit ¥205,000          205000
17356  Deposit ¥320,000          320000
17357  Deposit ¥195,000          195000

[15944 rows x 2 columns]


In [33]:
# Drop the original column
no_dup.drop("Deposit", axis = 1, inplace = True)

<br>
Key Price
<br>
Similar processing as the previous column.

In [34]:
# Cleaning the Key Price column

regex_extract("Key Price", "Key Price in Yen", r"\d+(.\d+)?", dtype = "int", yen = True)

no_dup["Key Payment Required?"] = np.where(no_dup["Key Price in Yen"] == 0, "No", "Yes")

                Key Price  Key Price in Yen
0            Key Money ¥0                 0
1            Key Money ¥0                 0
2            Key Money ¥0                 0
3            Key Money ¥0                 0
4            Key Money ¥0                 0
...                   ...               ...
17353        Key Money ¥0                 0
17354        Key Money ¥0                 0
17355  Key Money ¥205,000            205000
17356  Key Money ¥320,000            320000
17357  Key Money ¥195,000            195000

[15944 rows x 2 columns]


In [35]:
# Drop the original column
no_dup.drop("Key Price", axis = 1, inplace = True)

In [36]:
no_dup.head(2)


Floor       Year Built  \
0  Floor 2 / 8F  Year Built 2019   
1  Floor 1 / 4F  Year Built 2005   

                                     Nearest Station        Room Details  \
0  Nearest Station Kamikitazawa Station (5 min. w...  PRIVATE GUESTHOUSE   
1     Nearest Station Hatsudai Station (9 min. walk)        1K APARTMENT   

           City      Ward Negotiable  Monthly Rent Available Now?  \
0  SHIMOTAKAIDO  SUGINAMI         No      140000.0            Yes   
1      HOMMACHI   SHIBUYA        Yes      174000.0            Yes   

   Size in squared meters  Deposit in Yen Deposit Required?  Key Price in Yen  \
0                    12.0               0                No                 0   
1                    37.1               0                No                 0   

  Key Payment Required?  
0                    No  
1                    No

<br>
Floor
<br>
For this column I would have to devise my cleaning strategy after assessing the different values in this column. 

In [ ]:
# Seeing what are the possible values in this column
# Cleared output for legibility
no_dup.Floor.unique()

After assessing the possible values, I learned that I can simply extract all proceeding strings after "Floor", if the string "Floor" exists. If it doesn't exist, return "Not Available".

In [37]:
# Extracting floor information

floor_data = []
floor_pattern = r"(Floor) (.*)(F)"
floor = no_dup["Floor"].to_list()
for item in floor:
    try:
        match = re.search(floor_pattern,item).group(2)       # Find regex match based on the pattern
        floor_data.append(match)           # Appending the cleaned monthly rent 
    except:
        floor_data.append("Not Available")
    
# Updating dataframe with new data
no_dup["Floor Data"] = floor_data
# Qualitative check 
no_dup[["Floor", "Floor Data"]]

Floor Floor Data
0      Floor 2 / 8F      2 / 8
1      Floor 1 / 4F      1 / 4
2      Floor 5 / 8F      5 / 8
3      Floor 1 / 3F      1 / 3
4      Floor 1 / 3F      1 / 3
...             ...        ...
17353  Floor 1 / 2F      1 / 2
17354  Floor 1 / 5F      1 / 5
17355  Floor 7 / 7F      7 / 7
17356  Floor 7 / 7F      7 / 7
17357  Floor 2 / 3F      2 / 3

[15944 rows x 2 columns]

In [38]:
# Drop the original column
no_dup.drop("Floor", axis = 1, inplace = True)


<br>
Year Built
<br>
I will first scrape the year the building was built using regular expression. After the number of years are scraped, I will then create a new column that stores information about how old the building is, by subtracting the year built from 2021, the year that this dataset was scraped.

In [39]:
# Cleaning the Key Price column

regex_extract("Year Built", "Year", r"\d+(.\d+)?", dtype = "int", yen = False)

no_dup["Building Age"] = 2021 - no_dup["Year"]

            Year Built  Year
0      Year Built 2019  2019
1      Year Built 2005  2005
2      Year Built 2015  2015
3      Year Built 1994  1994
4      Year Built 2020  2020
...                ...   ...
17353  Year Built 2006  2006
17354  Year Built 2016  2016
17355  Year Built 2015  2015
17356  Year Built 2013  2013
17357  Year Built 2018  2018

[15944 rows x 2 columns]


In [40]:
# Drop the original column
no_dup.drop("Year Built", axis = 1, inplace = True)

<br>
Nearest Station
<br>
I will make two columns, one with nearest station name and another with how long of a walk it is to get to the station from the apartment. There is alot of variability in the values of this column, and one regex pattern may not be sufficient to match everything properly. Thus, I will do an interative process in which I will first separate the data between station name and how long it takes to walk there, and then I will use another regex pattern to extract the exact minutes required to walk there. There are also edge cases in which the listing describes taking a bus to the nearest station. I will account for those with a separate block of code.

In [41]:
# Cleaning the Nearest Station column

regex_pattern = r"Nearest Station (.*) \((.*)\)"
station_data = []
walk_data = []
orig_data = no_dup["Nearest Station"].to_list()
for item in orig_data:
    try:
        match1 = re.search(regex_pattern,item).group(1)       # Find regex match based on the pattern
        match2 = re.search(regex_pattern,item).group(2)       # Find regex match based on the pattern


        station_data.append(match1)                           # Appending the station date 
        walk_data.append(match2)                              # Appending the walk data
        
    except:                                                   # For edge cases when the data does not fit this type
        station_data.append(item)                             # Appending the original data for furthur cleaning.
        walk_data.append(item)
# Updating dataframe with new data
no_dup["Closest Station"] = station_data
no_dup["Walking Info"] = walk_data


# Qualitative check 
print(no_dup[["Nearest Station", "Closest Station"]])

                                         Nearest Station  \
0      Nearest Station Kamikitazawa Station (5 min. w...   
1         Nearest Station Hatsudai Station (9 min. walk)   
2          Nearest Station Yotsuya Station (7 min. walk)   
3         Nearest Station Hatsudai Station (9 min. walk)   
4         Nearest Station Asakusa Station (10 min. walk)   
...                                                  ...   
17353  Nearest Station Gakugei Daigaku Station (3 min...   
17354  Nearest Station Gakushuinshita Station (4 min....   
17355        Nearest Station Hiroo Station (9 min. walk)   
17356   Nearest Station Daikanyama Station (4 min. walk)   
17357    Nearest Station Tachikawa Station (7 min. walk)   

               Closest Station  
0         Kamikitazawa Station  
1             Hatsudai Station  
2              Yotsuya Station  
3             Hatsudai Station  
4              Asakusa Station  
...                        ...  
17353  Gakugei Daigaku Station  
17354   Gakushu

In [42]:
# Second round of regex matching to extract the stations that could have not been extracted before
pattern = r"Nearest Station.+ bus to (.+? ?.+? Station)"
final_station_data = []
for item in station_data:
    try:
        match1 = re.search(pattern,item).group(1)       # Find regex match based on the pattern

        final_station_data.append(match1)                     # Appending the station data
        
    except:                                                   # For edge cases when the data does not fit this type (for "Not Availables")
        final_station_data.append(item)                       # Appending the original data 

# Update dataframe
no_dup["Closest Station"] = final_station_data



In [43]:
# Extracting total time it takes to reach the nearest station, along with binary column of whether a bus is necessary

travel_time_data = []                                # Initializing data list that will be the new column
bus_data = []
for item in walk_data:
    if 'bus' in item:                                # Creating binary column of whether a bus is necessary to travel to the station
        bus_data.append("Yes")
    else:
        bus_data.append("No")
    
    minutes = re.findall(r"\d+", item)               # Looking for numbers in the walk data
    if not minutes:                                  # If no numbers were present in the data (Not Available)
        travel_time_data.append("Not Available")
    else:
        mins = []                                    # Create temporary list of numbers converted into numerical data
        for number in minutes:                       # Taking every number, converting to integer, and adding them together
            mins.append(int(number))
        travel_time_data.append(sum(mins))           # Appending total minutes it takes to get to the nearest station

            
# Update dataframe
no_dup["Time to Nearest Station"] = travel_time_data
no_dup["Bus Requried?"] = bus_data

# Drop the unnecessary column
no_dup.drop("Walking Info", axis = 1, inplace = True)

In [114]:
# Exporting to csv for easy manual checking using Excel
manual_check = no_dup[["Nearest Station", "Closest Station", "Walking Info", "Time to Nearest Station"]].to_csv(r"C:\Users\taked\Desktop\DataScience Projects\Non-Kaggle Datasets\Tokyo Apartments Project\manual2.csv")

In [44]:
# Drop the original column
no_dup.drop("Nearest Station", axis = 1, inplace = True)

In [45]:
# Final data
final_data = no_dup.copy()

In [50]:
# Saving the data
final_data.to_csv(r"C:\Users\taked\Desktop\DataScience Projects\Non-Kaggle Datasets\Tokyo Apartments Project\final_data.csv")

# Part 3 Loading the Data to Database (Load)

This is a quick, final part to close out the ETL process for the project. In part 1, I extracted the data from a data source (website), and in part 2 I transformed the data by cleaning the data with Pandas. In this final part of the script, I will be loading the data into a PostgreSQL server. 

In [51]:
# Importing required libraries
import psycopg2
from psycopg2.extensions import ISOLATION_LEVEL_AUTOCOMMIT
import pandas as pd
from sqlalchemy import create_engine 

In [52]:
# Connect to PostgreSQL
con = psycopg2.connect(
            host = "127.0.0.1",
            user = "postgres",
            password = "xxxxx" # Changed for security purposes
)

con.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)

# Cursor
cur = con.cursor()

In [53]:
# Create new database to house the Tokyo apartments dataset
cur.execute('DROP DATABASE IF EXISTS tokyo_apartments')
cur.execute('CREATE DATABASE tokyo_apartments')


In [54]:
# Close the cursor
cur.close()

# Close the connection
con.close()

In [55]:
# Reconnect to PostgreSQL, now directly to the database
con = psycopg2.connect(
            host = "127.0.0.1",
            database = "tokyo_apartments",
            user = "postgres",
            password = "xxxxx" # Changed for security purposes
)

con.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)

In [56]:
# Import csv to load into database
# Dropping the first column, as it is an index column
data = pd.read_csv(r"C:\Users\taked\Desktop\DataScience Projects\Non-Kaggle Datasets\Tokyo Apartments Project\final_data.csv").iloc[:,1:]

In [60]:
# Connect to db using SQLAlchemy to load Pandas dataframe to PostgreSQL

engine = create_engine('postgresql+psycopg2://postgres:xxxxx@127.0.0.1/tokyo_apartments')

In [61]:
# Load the dataframe to the database as a table with the name 'apartment_listings'
data.to_sql('apartment_listings', engine, if_exists = 'replace', index = False)

In [62]:
# Check if data has been stored properly
query = '''
SELECT *
FROM apartment_listings
LIMIT 10
'''


pd.read_sql_query(query, engine)

Room Details                      City        Ward Negotiable  \
0  PRIVATE GUESTHOUSE              SHIMOTAKAIDO    SUGINAMI         No   
1        1K APARTMENT                  HOMMACHI     SHIBUYA        Yes   
2      1LDK APARTMENT                    WAKABA    SHINJUKU        Yes   
3      1LDK APARTMENT                  HOMMACHI     SHIBUYA        Yes   
4        1R APARTMENT                   ASAKUSA       TAITO         No   
5        1K APARTMENT           SAKURASHIMMACHI    SETAGAYA         No   
6        1K APARTMENT  KITASHINAGAWA(1-4-CHOME)   SHINAGAWA         No   
7        1R APARTMENT                      CHUO      NAKANO         No   
8        1K APARTMENT                      AOTO  KATSUSHIKA         No   
9         3SLDK HOUSE            HIGASHIASAKUSA       TAITO        Yes   

   Monthly Rent Available Now?  Size in squared meters  Deposit in Yen  \
0      140000.0            Yes                   12.00               0   
1      174000.0            Yes                   37.10               0   
2      325000.0             No                   44.81               0   
3      177000.0            Yes                   57.00               0   
4      120000.0            Yes                   23.00          120000   
5       94000.0             No                   26.80           90000   
6      137000.0             No                   30.00          135000   
7       92000.0             No                   16.20           90000   
8       78000.0            Yes                   17.50               0   
9      270000.0            Yes                   87.00          270000   

  Deposit Required?  Key Price in Yen Key Payment Required? Floor Data  Year  \
0                No                 0                    No      2 / 8  2019   
1                No                 0                    No      1 / 4  2005   
2                No                 0                    No      5 / 8  2015   
3                No                 0                    No      1 / 3  1994   
4               Yes                 0                    No      1 / 3  2020   
5               Yes             90000                   Yes      6 / 6  1964   
6               Yes                 0                    No      3 / 4  1979   
7               Yes                 0                    No      1 / 2  1983   
8                No                 0                    No      2 / 4  1974   
9               Yes                 0                    No      1 / 1  2018   

   Building Age          Closest Station Time to Nearest Station Bus Requried?  
0             2     Kamikitazawa Station                       5            No  
1            16         Hatsudai Station                       9            No  
2             6          Yotsuya Station                       7            No  
3            27         Hatsudai Station                       9            No  
4             1          Asakusa Station                      10            No  
5            57  Sakurashinmachi Station                       2            No  
6            42    Kitashinagawa Station                       1            No  
7            38       Shinnakano Station                       4            No  
8            47          Kameari Station                      10            No  
9             3          Asakusa Station                      15            No

In [63]:
# Close the connections
engine.dispose()
con.close()

# Conclusion

With the preliminary transformation of the data finished, I am now ready to move onto doing analysis. I extracted the data from a website using a web scraper (E), and then I transformed the data using Python in a format that is ready to be analyzed (T), and finally stored the data in a database (L) Now, I will be using Tableau to analyze the data